### PR TITLE
feat: add CSV output format support with multiple CLI options

### LINKED ISSUE #392: Implement CSV output format
## Summary

Implement CSV (Comma-Separated Values) output format as the first phase of structured output formats outlined in #203. This will provide a human-readable, machine-parseable format that can be easily imported into spreadsheets and analytics tools.

## Background

This is a sub-issue of #203 (Support more output formats). The parent issue outlines a comprehensive plan for multiple output formats. This issue focuses specifically on implementing CSV format as Phase 1.

### Design Decisions
- **Human-readable focus**: The CSV format is primarily intended for human consumption and basic data export, not as a database import/export format
- **Use existing formatting**: Column values will use the same formatting as table output (no special CSV-specific formatting)
- **Standard library**: Use Go's `encoding/csv` package for RFC 4180 compliance

### 1. Add DisplayModeCSV enum value
```go
// In cli_output.go
const (
    // ... existing modes ...
    DisplayModeCSV  // Add after DisplayModeXML
)
```

### 2. Update CLI_FORMAT system variable
```go
// In system_variables.go - CLI_FORMAT setter
case "CSV":
    this.CLIFormat = DisplayModeCSV
```

### 3. Implement CSV output logic
```go
// In cli_output.go - printTableData()
case DisplayModeCSV:
    if len(columnNames) > 0 {
        csvWriter := csv.NewWriter(out)
        defer csvWriter.Flush()
        
        if \!sysVars.SkipColumnNames {
            if err := csvWriter.Write(columnNames); err \!= nil {
                slog.Error("csv.Writer.Write() failed for header", "err", err)
            }
        }
        for _, row := range result.Rows {
            if err := csvWriter.Write(row); err \!= nil {
                slog.Error("csv.Writer.Write() failed for row", "err", err)
            }
        }
    }
```

## Usage Example
```sql
-- Switch to CSV format
SET CLI_FORMAT = "CSV";

-- Query data
SELECT * FROM Users;

-- Output:
UserId,Name,Email
1,"John Doe","john@example.com"
2,"Jane Smith","jane@example.com"
```

## Testing Strategy
- Test CSV output with special characters (commas, quotes, newlines)
- Test with and without column names (CLI_SKIP_COLUMN_NAMES)
- Test empty result sets
- Verify RFC 4180 compliance through encoding/csv package
- Integration with existing output format tests

## Acceptance Criteria
- [ ] `SET CLI_FORMAT = "CSV";` successfully switches to CSV output
- [ ] CSV output properly escapes special characters via `encoding/csv`
- [ ] Header row respects `CLI_SKIP_COLUMN_NAMES` system variable
- [ ] All existing tests pass
- [ ] New CSV-specific tests pass

## Future Enhancements (Not in scope)
- Custom delimiter support (CLI_CSV_DELIMITER)
- Custom quote character (CLI_CSV_QUOTE_CHAR)
- NULL value representation options

These can be added in future iterations based on user feedback.

Parent issue: #203
EOF < /dev/null

### PR DESCRIPTION
**Pull request by apstndb**
## Summary

This PR implements CSV (Comma-Separated Values) output format support for query results as the first phase of implementing structured output formats requested in #203.

Fixes #392

## Changes

- Added `DisplayModeCSV` enum value to support CSV output format
- Implemented `printCSVTable` function using Go's `encoding/csv` package for RFC 4180 compliance
- Added `--csv` flag for easy CSV format selection (similar to `--html` and `--xml`)
- Added `--format` flag as a more generic option (accepts table/tab/vertical/html/xml/csv)
- Extracted common format parsing logic into `parseDisplayMode` function to avoid code duplication
- Updated `CLI_FORMAT` system variable to accept 'CSV' value (case-insensitive)
- Added comprehensive tests for CSV output including special character handling
- Updated documentation in README.md and system_variables.md

### Design Decisions

1. **Using `encoding/csv` package**: Instead of manual string escaping, we use Go's standard library `encoding/csv` package which automatically handles RFC 4180 compliant escaping of special characters (commas, quotes, newlines).

2. **Error Handling**: The `printCSVTable` function returns an error when there are no columns to output. This provides better error handling compared to silently producing empty output.

3. **Function Extraction**: CSV output logic was extracted into a dedicated `printCSVTable` function to ensure proper defer scoping for `csvWriter.Flush()`. This follows the same pattern as HTML and XML output functions.

4. **Human-Readable Focus**: As specified in the requirements, this CSV format is designed for human-readable output, not for database import/export operations. It uses the same column formatting as the TABLE format.

5. **Multiple CLI Options**: 
   - Added `--csv` flag for consistency with existing `--html` and `--xml` flags
   - Added `--format` flag inspired by `gcloud spanner databases execute-sql --format=` for a more generic approach
   - All format flags are mutually exclusive to prevent confusion

6. **Code Reuse**: Extracted `parseDisplayMode` function to unify format string parsing between `opts.Format` and `CLI_FORMAT` system variable, eliminating code duplication.

### Test Coverage

The implementation includes comprehensive tests covering:
- Simple CSV output
- Special characters (commas, quotes, newlines)
- Skip column names option
- Empty result handling
- Complex scenarios with mixed special characters
- Mutual exclusivity of all format flags

All tests pass successfully, ensuring robust handling of edge cases.

### Compatibility

- The CSV format is case-insensitive (accepts both 'CSV' and 'csv')
- Works with `CLI_SKIP_COLUMN_NAMES` system variable
- Compatible with all existing output control features
- Backward compatible with existing format flags

## Usage Examples

```bash
# Using the new --csv flag
spanner-mycli --csv -e "SELECT * FROM users"

# Using the generic --format flag
spanner-mycli --format=csv -e "SELECT * FROM users"

# Setting format via system variable
spanner-mycli -e "SET CLI_FORMAT = 'CSV'; SELECT * FROM users"
```

## Test Plan

- [x] Run `make check` - all tests pass
- [x] Manual testing with various SQL queries containing special characters
- [x] Verify `SET CLI_FORMAT = "CSV";` works correctly
- [x] Verify CSV output with `CLI_SKIP_COLUMN_NAMES = TRUE`
- [x] Verify error handling for queries with no results
- [x] Verify mutual exclusivity of format flags
- [x] Test new `--csv` and `--format` flags

🤖 Generated with [Claude Code](https://claude.ai/code)
EOF < /dev/null

### CODE REVIEW

File: cli_output.go [hunk 1]
```diff
@@ -660,3 +694,29 @@ func printXMLResultSet(out io.Writer, columnNames []string, rows []Row, skipColu
...
 	return err
 }
+
+// printCSVTable outputs query results in CSV format.
+// The format follows RFC 4180 with automatic escaping via encoding/csv.
+// Column headers are included unless skipColumnNames is true.
+func printCSVTable(out io.Writer, columnNames []string, rows []Row, skipColumnNames bool) error {
+	if len(columnNames) == 0 {
+		return fmt.Errorf("no columns to output")
+	}
```
[the remark this answers is not in the document]
Comment (apstndb) [reply cli_output.go:714]: Thank you for the review. I've addressed your concern by implementing a comprehensive solution:

1. **Added early return in `printTableData`**: When there are no columns, the function now returns early, preventing any output format functions from being called with empty columns.

2. **Unified error handling**: All format functions (HTML, XML, CSV) now consistently return an error for empty columns as a defensive programming measure.

3. **Added documentation**: Comments now explain that Spanner requires at least one column in SELECT queries, so empty columns should only occur in edge cases.

4. **Updated tests**: All tests now expect errors when format functions are called directly with empty columns.

This approach ensures consistent behavior across all output formats while maintaining the defensive programming practice of checking for edge cases.
EOF < /dev/null

Fixed in commit aa96358.

File: cli_output.go [hunk 2]
```diff
@@ -65,6 +93,12 @@ func printTableData(sysVars *systemVariables, screenWidth int, out io.Writer, re
 
 	columnNames := renderTableHeader(result.TableHeader, false)
 
+	// Early return if no columns - Spanner requires at least one column in SELECT,
+	// so this only happens for edge cases where no output is expected
+	if len(columnNames) == 0 {
+		return
+	}
```
[the remark this answers is not in the document]
Comment (apstndb) [reply cli_output.go]: Thank you for the thorough review. You're absolutely right - the early return does prevent the error logging in individual format functions from being reached.

After analyzing the code, I found that:
1. The early return is actually preventing a panic in Vertical mode (line 178) that would occur if a row has more columns than headers
2. Other display modes (TABLE, TAB) already handle empty columnNames gracefully

I've kept the early return but added warning logging as suggested in the original feedback, and fixed the Vertical mode panic issue by adding proper bounds checking. This approach:
- Maintains the defensive early return that prevents panics
- Makes the edge case visible through logging when rows exist but headers don't
- Fixes the actual bug in Vertical mode
- Preserves the error returns in format functions for direct calls (e.g., in tests)

Fixed in commit 0271cd7.
EOF < /dev/null

Fixed in commit 0271cd7.

File: main.go [hunk 3]
```diff
@@ -82,6 +82,8 @@ type spannerOptions struct {
 	Table                     bool              `long:"table" short:"t" description:"Display output in table format for batch mode." default-mask:"-"`
 	HTML                      bool              `long:"html" description:"Display output in HTML format." default-mask:"-"`
 	XML                       bool              `long:"xml" description:"Display output in XML format." default-mask:"-"`
+	CSV                       bool              `long:"csv" description:"Display output in CSV format." default-mask:"-"`
+	Format                    string            `long:"format" description:"Output format (table, tab, vertical, html, xml, csv)" choice:"table" choice:"tab" choice:"vertical" choice:"html" choice:"xml" choice:"csv" default-mask:"-"`
```
[the remark this answers is not in the document]
Comment (apstndb) [reply main.go]: Excellent suggestion\! I've implemented your recommendations:

1. Removed the `choice` tags from the Format field definition
2. Added validation in ValidateSpannerOptions using the parseDisplayMode function
3. This now provides case-insensitive format validation consistent with CLI_FORMAT

The --format flag now accepts format names in any case (e.g., `--format=CSV`, `--format=csv`, `--format=CsV` all work), providing a better user experience.

Fixed in commit 0271cd7.
EOF < /dev/null

Fixed in commit 0271cd7.

File: cli_output.go [hunk 4]
```diff
@@ -660,3 +711,33 @@ func printXMLResultSet(out io.Writer, columnNames []string, rows []Row, skipColu
...
+
+	if !skipColumnNames {
+		if err := csvWriter.Write(columnNames); err != nil {
+			return err
+		}
+	}
+
+	for _, row := range rows {
+		if err := csvWriter.Write(row); err != nil {
+			return err
```
[the remark this answers is not in the document]
Comment (apstndb) [reply cli_output.go:739]: Fixed by adding explicit flush with error checking to prevent data loss when writing CSV output

Fixed in commit 625deb6.

File: cli_output.go [hunk 5]
```diff
@@ -660,3 +711,38 @@ func printXMLResultSet(out io.Writer, columnNames []string, rows []Row, skipColu
...
+		}
+	}
+
+	// Explicitly flush to ensure all data is written
+	csvWriter.Flush()
+	if err := csvWriter.Error(); err != nil {
+		return fmt.Errorf("csv.Writer.Flush() failed: %w", err)
+	}
+
+	return nil
```
[the remark this answers is not in the document]
Comment (apstndb) [reply cli_output.go]: Fixed by using defer statement for csvWriter.Flush() to ensure buffer is always flushed even on error, following idiomatic Go pattern

Fixed in commit 3ba7338.
